### PR TITLE
Add support for PDB with the teleport-cluster helm chart

### PR DESCRIPTION
**Pull request by stefansedich**
This PR adds support for a PDB as part of the teleport-cluster chart, for the sole reason that if we truly want to support HA then we should likely configure a PDB.

I have disabled it by default and have made it an opt-in thing, we can also enable it when `replicaCount>1` so am open to suggestions.

### CODE REVIEW

File: examples/chart/teleport-cluster/templates/pdb.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{{- if .Values.podDisruptionBudget.enabled }}
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ .Release.Name }}
+  namespace: {{ .Release.Namespace }}
```
Comment (stefansedich) [inline examples/chart/teleport-cluster/templates/pdb.yaml:6]: It is normally frowned upon to set the namespace in chart resources but I have done this as it is already done everywhere else.

File: examples/chart/teleport-cluster/templates/pdb.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{{- if .Values.podDisruptionBudget.enabled }}
+apiVersion: policy/v1beta1
```
Comment (stefansedich) [inline examples/chart/teleport-cluster/templates/pdb.yaml:2]: 1.21 introduced `policy/v1` but I have kept v1beta1 so this chart can support earlier k8s versions.

**Comment by russjones** [comment]
@webvictim Can you review this?

File: examples/chart/teleport-cluster/values.yaml [hunk 3]
```diff
@@ -35,6 +35,12 @@ enterprise: false
 podSecurityPolicy:
   enabled: true
 
+# If enabled will create a Pod Disruption Budget
+# https://kubernetes.io/docs/concepts/workloads/pods/disruptions/
+podDisruptionBudget:
+  enabled: false
+  minAvailable: 1
```
Comment (webvictim) [inline examples/chart/teleport-cluster/values.yaml:42]: I think that this should be under the `highAvailability` section as it's only relevant for that type of deployment.

**Review by webvictim — CHANGES_REQUESTED** [review]
Added some feedback.

Also, please add this value to the Helm chart reference at https://github.com/gravitational/teleport/blob/master/docs/pages/kubernetes-access/helm/reference.mdx so it's kept up to date.

**Comment by stefansedich** [comment]
> Added some feedback.
> 
> Also, please add this value to the Helm chart reference at https://github.com/gravitational/teleport/blob/master/docs/pages/kubernetes-access/helm/reference.mdx so it's kept up to date.

@webvictim feedback implemented and added to the docs.

File: examples/chart/teleport-cluster/values.schema.json [hunk 4]
```diff
@@ -199,7 +200,27 @@
...
+                            "type": "boolean",
+                            "default": false
+                        },
+                        "minAvailable": {
+                            "$id": "#/properties/highAvailability/properties/podDisruptionBudget/properties/minAvailable",
+                            "type": "integer",
+                            "default": 1
+                        }
+                    }
+                },
```
Comment (webvictim) [inline examples/chart/teleport-cluster/values.schema.json:223]:
```suggestion
                }
```

Tests are failing, I think this is probably why.

**Comment by webvictim** [comment]
@stefansedich Thanks! One final thing - could you please add a sample values file `examples/chart/teleport-cluster/.lint/pdb.yaml` to cover this new addition which will ensure that the new template gets automatically evaluated on PRs?

See something like https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-cluster/.lint/volumes.yaml for an example.

File: docs/pages/kubernetes-access/helm/reference.mdx [hunk 5]
```diff
@@ -400,6 +400,58 @@ Teleport pods must not be scheduled on the same physical host.
...
 </Tabs>
 
+## `highAvailability.podDisruptionBudget`
+
+### `highAvailability.podDisruptionBudget.enabled`
+
+| Type | Default value | Can be used in `custom` mode? |
+| - | - | - |
+| `bool` | `false` | ✅ (when using HA storage) |
+
```
Comment (inertial-frame) [inline docs/pages/kubernetes-access/helm/reference.mdx:410]: Docs-wise this is great thanks for the helpful reference update!

**Comment by stefansedich** [comment]
> @stefansedich Thanks! One final thing - could you please add a sample values file `examples/chart/teleport-cluster/.lint/pdb.yaml` to cover this new addition which will ensure that the new template gets automatically evaluated on PRs?
> 
> See something like https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-cluster/.lint/volumes.yaml for an example.

@webvictim done, hopefully I got that right :smile:

**Comment by webvictim** [comment]
@stefansedich Yep, looks good. Let's see what the linter says.

File: examples/chart/teleport-cluster/values.schema.json (hunk 4 above)
Comment (webvictim) [inline examples/chart/teleport-cluster/values.schema.json:223]: Linter is unhappy with this: https://drone.teleport.dev/gravitational/teleport/476/1/5

```suggestion
                }
```

Comment (stefansedich) [reply examples/chart/teleport-cluster/values.schema.json]: :facepalm: my bad!

File: examples/chart/teleport-cluster/.lint/pdb.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+clusterName: helm-lint
+highAvailability:
```
Comment (webvictim) [inline examples/chart/teleport-cluster/.lint/pdb.yaml:2]: Linter is throwing another warning:

```
1 chart(s) linted, 0 chart(s) failed
--
241 | Using values from 'examples/chart/teleport-cluster/.lint/pdb.yaml': ==> Linting examples/chart/teleport-cluster
242 | [ERROR] templates/: template: teleport-cluster/templates/deployment.yaml:5:4: executing "teleport-cluster/templates/deployment.yaml" at <fail "Cannot enable multiple replicas in standalone mode, use a different chartMode which supports high availability - see README and docs">: error calling fail: Cannot enable multiple replicas in standalone mode, use a different chartMode which supports high availability - see README and docs

```

I think adding `chartMode: custom` in here will resolve this issue.

Comment (stefansedich) [reply examples/chart/teleport-cluster/.lint/pdb.yaml:3]: Lets see if that does the trick.

**Comment by stefansedich** [comment]
@webvictim what was needed now to get this in?

**Review by webvictim — APPROVED** [review]

**Comment by webvictim** [comment]
Looks fine now. @klizhentas @r0mant @russjones Please codeowner review.

**Comment by stefansedich** [comment]
@webvictim @klizhentas @r0mant @russjones any chance of getting this one in?

**Comment by webvictim** [comment]
This is fine by me, just needs a rubber stamp from @russjones @klizhentas or @r0mant.

**Review by russjones — APPROVED** [review]
Bot.

**Comment by russjones** [comment]
@webvictim Please backport this to `branch/v7` as well, that's where we will be releasing Helm charts from soon.

**Pull request merged**